Prometheus metrics

ysf3rdm opened the pull request:
**Changes**

This only implements user account metrics for now as outlined in #3016. This PR adds a new `UserMetrics` class that tracks user accounts, permissions, authentication providers, activity, and security metrics.

**Implemented Metrics:**

- [x] `jellyfin_user_accounts_total` - Total number of user accounts
- [x] `jellyfin_user_accounts_active` - Number of active (non-disabled) accounts
- [x] `jellyfin_user_accounts_admin` - Number of administrator accounts
- [x] `jellyfin_users_by_permission` - Users with specific permissions (labeled by permission type)
- [x] `jellyfin_users_by_auth_provider` - Users by authentication provider (labeled by provider)
- [x] `jellyfin_users_external_auth_total` - Total users using external authentication
- [x] `jellyfin_users_recently_active` - Users active within last 30 days
- [x] `jellyfin_users_failed_logins` - Users with failed login attempts (labeled by attempt range)

**Planned Metrics (from #3016):**

- [x] Session metrics (active sessions, device types, client types)
- [x] Stream metrics (play methods, transcoding, codecs)
- [ ] Library metrics (library counts, media types, codec distribution)

**Questions for Reviewers**

1. **Label Cardinality**: Currently using `AuthenticationProviderId` as a dynamic label for the auth provider metric. In typical Jellyfin installations, how many distinct auth providers exist? Should this be categorized into "Default" vs "External" instead to prevent unbounded cardinality growth?

2. **Architecture**: Should the remaining metrics (sessions, streams, libraries) be in separate classes following the same pattern, or would you prefer a different organization?

3. **Additional Metrics**: Are there any other user-related or Jellyfin-specific metrics you'd like to see added? Please suggest in the comments.

**Issues**

#3016

ysf3rdm committed:
User Metrics Added

ysf3rdm committed:
circular dependency issue resolved

ysf3rdm commented:
When new features are accepted, should I wait for version 10.11.1 or 10.12?

JPVenson commented on Jellyfin.Server.Implementations/Users/UserManager.cs:
```diff
@@ -73,7 +76,8 @@ public UserManager(
             IEnumerable<IPasswordResetProvider> passwordResetProviders,
-            IEnumerable<IAuthenticationProvider> authenticationProviders)
+            IEnumerable<IAuthenticationProvider> authenticationProviders,
+            UserMetrics? userMetrics = null)
```
i am not too happy with this dependecy. I think it would be far better if we invert the responsibilities here or at the very least make it far more generic so that other code could also benefit from it

JPVenson reviewed (changes requested):
Aren't you missing the actual pipeline registration to enable the prometheus endpoint?
https://github.com/prometheus-net/prometheus-net#aspnet-core-exporter-middleware

ysf3rdm committed:
Merge branch 'jellyfin:master' into prometheus-metrics-10.11

renovate[bot] committed:
Update appleboy/ssh-action action to v1.2.3

dpieski committed:
Update Jellyfin Server version in issue template (#15398)

renovate[bot] committed:
Update dependency z440.atl.core to 7.7.0

renovate[bot] committed:
Update dependency FsCheck.Xunit to 3.3.2

renovate[bot] committed:
Update dependency dotnet-ef to v9.0.11

renovate[bot] committed:
Update Microsoft

Tminus10official committed:
Translated using Weblate (English (Pirate))
Translation: Jellyfin/Jellyfin
Translate-URL: https://translate.jellyfin.org/projects/jellyfin/jellyfin-core/en@pirate/

renovate[bot] committed:
Update CI dependencies

hoanghuy309 committed:
Translated using Weblate (Vietnamese)
Translation: Jellyfin/Jellyfin
Translate-URL: https://translate.jellyfin.org/projects/jellyfin/jellyfin-core/vi/

renovate[bot] committed:
Update dependency z440.atl.core to 7.8.0

renovate[bot] committed:
Update actions/checkout action to v5.0.1

theguymadmax committed:
Backport pull request #14955 from jellyfin/release-10.11.z

Fix tmdbid not detected in single movie folder

Original-merge: def5956cd1afe8848c0e232fa477720c4158832f

Merged-by: crobibero <cody@robibe.ro>

Backported-by: Bond_009 <bond.009@outlook.com>

revam committed:
Backport pull request #15263 from jellyfin/release-10.11.z

Resolve symlinks for static media source infos

Original-merge: 3b2d64995aab63ebaa6832c059a3cc0bdebe90dc

Merged-by: crobibero <cody@robibe.ro>

Backported-by: Bond_009 <bond.009@outlook.com>

evanreichard committed:
Backport pull request #15369 from jellyfin/release-10.11.z

feat(sqlite): add timeout config

Original-merge: c2e5081d64e519a74d47df23335bb228fea8ec7e

Merged-by: crobibero <cody@robibe.ro>

Backported-by: Bond_009 <bond.009@outlook.com>

theguymadmax committed:
Backport pull request #15373 from jellyfin/release-10.11.z

Fix collection grouping in mixed libraries

Original-merge: 13c4517a66e3f857cc4acc9b2fa3505297d554eb

Merged-by: crobibero <cody@robibe.ro>

Backported-by: Bond_009 <bond.009@outlook.com>

theguymadmax committed:
Backport pull request #15380 from jellyfin/release-10.11.z

Fix item count display for collapsed items

Original-merge: 8f71922734d42591b3236f4c52d9692f1b191da2

Merged-by: crobibero <cody@robibe.ro>

Backported-by: Bond_009 <bond.009@outlook.com>

theguymadmax committed:
Backport pull request #15381 from jellyfin/release-10.11.z

Fix name filters to use only SortName

Original-merge: 7222910b05dff772fad22d4f557fad20578fa275

Merged-by: crobibero <cody@robibe.ro>

Backported-by: Bond_009 <bond.009@outlook.com>

Bond-009 committed:
Backport pull request #15390 from jellyfin/release-10.11.z

Don't enforce a minimum amount of free space for the tmp and log dirs

Original-merge: 097cb87f6f6df662361a4cd536b56470e4cd68a3

Merged-by: crobibero <cody@robibe.ro>

Backported-by: Bond_009 <bond.009@outlook.com>

theguymadmax committed:
Backport pull request #15404 from jellyfin/release-10.11.z

Improve season folder parsing

Original-merge: 2e5ced50986c37b19b5f4ef34d730fc56a51535a

Merged-by: crobibero <cody@robibe.ro>

Backported-by: Bond_009 <bond.009@outlook.com>

JPVenson committed:
Backport pull request #15413 from jellyfin/release-10.11.z

Fixed missing sort argument

Original-merge: 91c3b1617e06283c88f36bc63046b99c993cb774

Merged-by: crobibero <cody@robibe.ro>

Backported-by: Bond_009 <bond.009@outlook.com>

nielsvanvelzen committed:
Backport pull request #15422 from jellyfin/release-10.11.z

Update branding in Swagger page

Original-merge: d1406302085148aea4d3b03dab78c4d49909c4fe

Merged-by: crobibero <cody@robibe.ro>

Backported-by: Bond_009 <bond.009@outlook.com>

theguymadmax committed:
Backport pull request #15423 from jellyfin/release-10.11.z

Invalidate parent folder's cache on deletion/creation

Original-merge: 49efd68fc7ef4b70b38151a177502bbdb3adede0

Merged-by: crobibero <cody@robibe.ro>

Backported-by: Bond_009 <bond.009@outlook.com>

theguymadmax committed:
Backport pull request #15435 from jellyfin/release-10.11.z

Fix search terms using diacritics

Original-merge: 63a3e552978010e905ea9b4e258c0c2e153ddfec

Merged-by: crobibero <cody@robibe.ro>

Backported-by: Bond_009 <bond.009@outlook.com>

IceStormNG committed:
Backport pull request #15441 from jellyfin/release-10.11.z

Fix System.NullReferenceException when people's role is null (10.11.z)

Original-merge: 5a9a8363f45ed34b2db1f8f59c2be69ff458f8ab

Merged-by: Bond-009 <bond.009@outlook.com>

Backported-by: Bond_009 <bond.009@outlook.com>

theguymadmax committed:
Backport pull request #15446 from jellyfin/release-10.11.z

Fix AncestorIds not migrating

Original-merge: 177b6464ca1b6772a191dbf9c5595708f91fc0fa

Merged-by: crobibero <cody@robibe.ro>

Backported-by: Bond_009 <bond.009@outlook.com>

theguymadmax committed:
Backport pull request #15462 from jellyfin/release-10.11.z

Fix NullReferenceException in GetPathProtocol when path is null

Original-merge: 7c1063177f5647e07c634d36694f0bdec8fe4ff1

Merged-by: joshuaboniface <joshua@boniface.me>

Backported-by: Bond_009 <bond.009@outlook.com>

Bond-009 committed:
Backport pull request #15466 from jellyfin/release-10.11.z

Don't error out when searching for marker files fails

Original-merge: f4a846aa4dcffb3be7b701f806b24cb8dd6b7c5d

Merged-by: crobibero <cody@robibe.ro>

Backported-by: Bond_009 <bond.009@outlook.com>

Bond-009 committed:
Backport pull request #15468 from jellyfin/release-10.11.z

Check if target exists before trying to follow it

Original-merge: 5878b1ffc569bc7e7204a07e77e0bb57d3984e56

Merged-by: joshuaboniface <joshua@boniface.me>

Backported-by: Bond_009 <bond.009@outlook.com>

theguymadmax committed:
Backport pull request #15472 from jellyfin/release-10.11.z

Fix series DateLastMediaAdded not updating when new episodes are added

Original-merge: abfbaca33686ab5214182992ab644a6e24d4b180

Merged-by: crobibero <cody@robibe.ro>

Backported-by: Bond_009 <bond.009@outlook.com>

CBPJ committed:
Backport pull request #15487 from jellyfin/release-10.11.z

Fix gitignore-style not working properly on windows.

Original-merge: 435bb14bb266916e9c6f100c4324a94c36126e06

Merged-by: crobibero <cody@robibe.ro>

Backported-by: Bond_009 <bond.009@outlook.com>

theguymadmax committed:
Backport pull request #15493 from jellyfin/release-10.11.z

Remove InheritedTags and update tag filtering logic

Original-merge: 4b38e35bbbb65c77f251288ff64ee28da4a48943

Merged-by: crobibero <cody@robibe.ro>

Backported-by: Bond_009 <bond.009@outlook.com>

theguymadmax committed:
Backport pull request #15501 from jellyfin/release-10.11.z

Fix .ignore handling for directories

Original-merge: e8150428b62668e062a3432960f98684d3b352cb

Merged-by: crobibero <cody@robibe.ro>

Backported-by: Bond_009 <bond.009@outlook.com>

Iksas committed:
Backport pull request #15502 from jellyfin/release-10.11.z

Fix font extraction for certain transcoding settings

Original-merge: ee34c75386cc1a0ca0e15196a43b685fa0e73130

Merged-by: crobibero <cody@robibe.ro>

Backported-by: Bond_009 <bond.009@outlook.com>

theguymadmax committed:
Backport pull request #15508 from jellyfin/release-10.11.z

Fix playlist DateCreated and DateLastMediaAdded not being set

Original-merge: 078f9584ed3622eed3516488026cbb6e42242bba

Merged-by: crobibero <cody@robibe.ro>

Backported-by: Bond_009 <bond.009@outlook.com>

theguymadmax committed:
Backport pull request #15514 from jellyfin/release-10.11.z

Add 1 minute tolerance for NFO change detection

Original-merge: 6566188e453b42604dbb3ce532937951e88565d0

Merged-by: crobibero <cody@robibe.ro>

Backported-by: Bond_009 <bond.009@outlook.com>

theguymadmax committed:
Backport pull request #15529 from jellyfin/release-10.11.z

Fix movie titles using folder name when NFO saver is enabled

Original-merge: f8e012582a8819d18ad933fd65eade936bdc946d

Merged-by: crobibero <cody@robibe.ro>

Backported-by: Bond_009 <bond.009@outlook.com>

Rufis72 committed:
Translated using Weblate (English (Pirate))
Translation: Jellyfin/Jellyfin
Translate-URL: https://translate.jellyfin.org/projects/jellyfin/jellyfin-core/en@pirate/

renovate[bot] committed:
Update github/codeql-action action to v4.31.4

anthonylavado committed:
Add all 10.11 versions to issue template (#15565)

renovate[bot] committed:
Update actions/checkout action to v6

oddstr13 commented on Jellyfin.Server.Implementations/Metrics/UserMetrics.cs:
```diff
@@ -0,0 +1,249 @@
+using Microsoft.Extensions.Logging;
+using Prometheus;
+
+namespace Jellyfin.Server.Implementations.Metrics
```
Would that be `Jellyfin.Server.Implementations.Metrics.UserMetrics` in this case?

How is the general structure of the metrics part of this PR, is this how you'd want it structured?

I've been wanting to try my hand at adding some metrics too, but figuring out how and where to place it has been daunting enough that I've just kicked that can down the road. If this structure looks fine other than a few nits, then it looks like a easy starting point for adding more metrics.

ysf3rdm replied:
Thanks for the feedback. I’ve updated the structure based on the comments I received and I think the current version is much better, but I haven’t had the time to push the commits yet. 

To remove the dependencies of the metric classes from other classes — for example, removing UserMetrics’ dependency inside UserManager — using events to update the metrics seemed like the most reasonable approach, and that’s the direction I’m planning to take. 

I also think categorizing the metrics into separate classes makes the overall structure clearer and makes it easier to know exactly where a new metric should go. Once the main structure is finalized, adding new metrics should be as easy as you mentioned.

renovate[bot] committed:
Update dependency z440.atl.core to 7.9.0

rimasx committed:
Translated using Weblate (Estonian)
Translation: Jellyfin/Jellyfin
Translate-URL: https://translate.jellyfin.org/projects/jellyfin/jellyfin-core/et/

renovate[bot] committed:
Update github/codeql-action action to v4.31.5

ysf3rdm committed:
Session Metrics added, UpdateMetrics use related events and metric service registration done via HostedService

ysf3rdm committed:
Merge branch 'jellyfin:master' into prometheus-metrics-10.11

jpds commented on Jellyfin.Server.Implementations/Metrics/SessionMetrics.cs:
```diff
@@ -0,0 +1,184 @@
+            .CreateGauge("jellyfin_sessions_active_total", "Number of currently active sessions");
+
+        private static readonly Gauge _totalSessions = Prometheus.Metrics
+            .CreateGauge("jellyfin_sessions_total", "Total number of sessions, active and inactive");
```
I don't think there's a need for two different metric names when you can consolidate this into `jellyfin_sessions_total` with a `state={active,inactive}` label.

jpds reviewed (commented):
Awesome PR. Not a Jellyfin developer - but long-time Prometheus user.

jpds commented on Jellyfin.Server.Implementations/Metrics/UserMetrics.cs:
```diff
@@ -0,0 +1,281 @@
+            .CreateGauge("jellyfin_users_external_auth_total", "Total number of users using external authentication");
+
+        private static readonly Gauge _recentlyActiveUsers = Prometheus.Metrics
+            .CreateGauge("jellyfin_users_recently_active", "Number of users active within the last 30 days");
```
This is also not needed as one can just do: `sum(jellyfin_sessions_total{state="active"})`
